src/Perpetuum/Zones/Effects/ZoneEffects/ZoneEffectHandler.cs

```diff
@@ -1,22 +1,85 @@
-﻿using Perpetuum.Players;
+﻿using Perpetuum.Log;
+using Perpetuum.Players;
+using Perpetuum.Units;
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
 
 namespace Perpetuum.Zones.Effects.ZoneEffects
 {
     public class ZoneEffectHandler : IZoneEffectHandler
     {
-        private readonly IZoneEffectRepository _zoneEffectRepository;
-        public ZoneEffectHandler(IZoneEffectRepository zoneEffectRepository)
+        private readonly IZone _zone;
+        private readonly ConcurrentDictionary<ZoneEffect, byte> _effects;
+
+        public ZoneEffectHandler(IZone zone)
+        {
+            _zone = zone;
+            _effects = new ConcurrentDictionary<ZoneEffect, byte>();
+            InitCollection();
+        }
+
+        private void InitCollection()
+        {
+            foreach (var zoneEffect in ZoneEffectReader.GetStaticZoneEffects(_zone))
+            {
+                _effects.Add(zoneEffect, byte.MinValue);
+            }
+        }
+
+        public void AddEffect(ZoneEffect zoneEffect)
+        {
+            _effects.TryAdd(zoneEffect, byte.MinValue);
+            OnZoneEffectAdded(zoneEffect);
+        }
+
+        public void RemoveEffect(ZoneEffect zoneEffect)
+        {
+            _effects.TryRemove(zoneEffect, out byte b);
+            OnZoneEffectRemoved(zoneEffect);
+        }
+
+        private void OnZoneEffectRemoved(ZoneEffect effect)
+        {
+            var eligibleUnits = _zone.Units.Where(u => u.EffectHandler.ContainsEffect(effect.Effect));
+            foreach (var unit in eligibleUnits)
+            {
+                unit.EffectHandler.RemoveEffectsByType(effect.Effect);
+                Logger.DebugInfo($"Removing {effect.Effect} to {unit} on zone:{_zone.Id}");
+            }
+        }
+
+        private void OnZoneEffectAdded(ZoneEffect effect)
+        {
+            var eligibleUnits = _zone.Units.Where(u => CanApplyEffect(effect)(u));
+            foreach (var unit in eligibleUnits)
+            {
+                var builder = unit.NewEffectBuilder().SetType(effect.Effect).SetOwnerToSource();
+                unit.ApplyEffect(builder);
+                Logger.DebugInfo($"Adding {effect.Effect} to {unit} on zone:{_zone.Id}");
+            }
+        }
+
+        private IEnumerable<ZoneEffect> GetEffects()
+        {
+            return _effects.Keys;
+        }
+
+        private Func<Unit, bool> CanApplyEffect(ZoneEffect effect)
         {
-            _zoneEffectRepository = zoneEffectRepository;
+            return u => !u.EffectHandler.ContainsEffect(effect.Effect) && (!effect.PlayerOnly || (effect.PlayerOnly && u is Player));
```

Comment:
`return u => !u.EffectHandler.ContainsEffect(effect.Effect) && (!effect.PlayerOnly || u is Player));` should be the same difference no?
`(!effect.PlayerOnly || (effect.PlayerOnly && u is Player))` If `PlayerOnly` is false, the statement returns true. If we get to the `or` part of the statement, `PlayerOnly` can *only* be true.

Reply:
oh you are right, should have written out a truth table XD  good catch